### PR TITLE
Issue #13999: Resolve Pitest AbstractJavadocCheck mutation-1 suppression

### PR DESCRIPTION
**Pull request by chris-carneiro**
Part of Issue #13999
 
[AbstractJavadocCheck Documentation](https://checkstyle.org/apidocs/com/puppycrawl/tools/checkstyle/checks/javadoc/AbstractJavadocCheck.html)

Mutation - 1
https://github.com/checkstyle/checkstyle/blob/42f3c4f33790050b75341b6ad1c5afe0083e762c/config/pitest-suppressions/pitest-javadoc-suppressions.xml#L3-L10

Caching by column was removed , see reason at https://github.com/checkstyle/checkstyle/pull/15798#issuecomment-2429262257 and [why it works ](https://github.com/checkstyle/checkstyle/pull/15798#discussion_r1814893199)

### CODE REVIEW

**Comment by romani** [comment]
You need to remove suppression, https://github.com/checkstyle/checkstyle/actions/runs/11455330207/job/31871144519?pr=15798

But if you killed muratation by code modification. You probably can extend some test to validate column in violation message. Can we?
I think we have at least one Check that reports violation on column.

**Comment by romani** [comment]
yes, you are right, it is cache of block Comment Nodes.
So we need two javadoc comments on single line, `/** Summary */ /** some text.*/`.

This is not realistic code, nobody writes like this, and javadoc tool from jdk will not consider it at all, so it is not javadoc comment actually.

```
$ cat Test.java 
/** Summary */ /** some text.*/
public class Test {
}

$ javadoc Test.java
Loading source file Test.java...
Constructing Javadoc information...
Standard Doclet version 11.0.14.1
Building tree for all the packages and classes...
Generating ./Test.html...
Generating ./package-summary.html...
Generating ./package-tree.html...
Generating ./constant-values.html...
Building index for all the packages and classes...
Generating ./overview-tree.html...
Generating ./index-all.html...
Building index for all classes...
Generating ./allclasses-index.html...
Generating ./allpackages-index.html...
Generating ./deprecated-list.html...
Building index for all classes...
Generating ./allclasses.html...
Generating ./allclasses.html...
Generating ./index.html...
Generating ./help-doc.html...

```

you can see that there is no "Summary"
image


so it means that `if (JavadocUtil.isJavadocComment(blockCommentNode)) {` is not correct, as it is too simple in implementation for detection of javadoc.

`/** Summary */ ` is not a javadoc comment, it is block block comment with javadoc identifier.
https://checkstyle.org/writingjavadocchecks.html#What_is_Javadoc_comment

So it means that mutation is correct, and we by nature of javadoc can not have two comments on same line, so defining column is over-caution.
So we need ` final int treeCacheKey = blockCommentNode.getLineNo();` and slight refactoring on this.

**Comment by romani** [comment]
Checker is not happy https://github.com/checkstyle/checkstyle/actions/runs/11462075960/job/31892618248?pr=15798#step:6:1705

Can we fix it without suppression?

**Comment by romani** [comment]
To fix it please apply patch https://github.com/checkstyle/checkstyle/actions/runs/11462075960/job/31892618248?pr=15798#step:7:8

Suppressed message is changed.

**Comment by chris-carneiro** [comment]
> Can we fix it without suppression?

I'm gonna try

**Comment by romani** [comment]
Xwiki failure is not related.

**Comment by chris-carneiro** [comment]
> Xwiki failure is not related.

Right, seems like both failures are unrelated to the updates.
"check_issues" i understand is a false positive, the "no-error-xwiki" failure, I'll check tomorrow if I can do anything about it.

**Comment by romani** [comment]
I have worries about cache, and two comments on same line, it means that we will use first comment in line but actually second is considered as javadoc. So checks will skip being executed on second.

That is not good.
We can not miss violations on second comment.

**Comment by chris-carneiro** [comment]
> I have worries about cache, and two comments on same line, it means that we will use first comment in line but actually second is considered as javadoc. So checks will skip being executed on second.
> 
> That is not good. We can not miss violations on second comment.

I have addressed your concerns by adding two test cases based on my interpretation of the issue you raised. I ran both implementations (the current release and my PR) and examined the Javadoc tree structures for each.

In both implementations, I found no differences in the tree structures produced by my tests. I've documented my process and results in the following Gist:

[JavadocTreeStructures](https://gist.github.com/chris-carneiro/691d92082c40539376a9ef135f04f5f4)

To further illustrate the results, I've included a diff test output comparing the two versions. The files prefixed with current_ correspond to the released version, while those prefixed with 13999_ use a custom build that includes the updates from this PR.

Here's the visual output from the diff test for your reference:
Screenshot 2024-10-23 at 14 46 18

@romani what do you think?

**Comment by romani** [comment]
Github, generate report for JavadocBlockTagLocation/all-examples-in-one

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/abstractjavadoc/InputAbstractJavadocCache4.java
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/abstractjavadoc/InputAbstractJavadocCache4.java:14]: This is expected behavior. Good.
But need to look closer to code explain why.
As I think we will not parse second comment as it is in cache after first comment is parsed

**Review by romani — CHANGES_REQUESTED** [review]
Rebase will fix CIs.
I do red sign just to double check. I miss something.

**Comment by romani** [comment]
Please squash all commits in one.

**Comment by chris-carneiro** [comment]
> Please squash all commits in one.

Oops  my bad

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/abstractjavadoc/InputAbstractJavadocCache4.java
Comment (chris-carneiro) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/abstractjavadoc/InputAbstractJavadocCache4.java:14]: I understand your point, I'm willing to give it a try and understand what's going on. 
I still have other stuff to work on though.

Do you think it is worth the time investment ? Or maybe the risk to apply this update is too high for the benefit ?

Comment (romani) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/abstractjavadoc/InputAbstractJavadocCache4.java:14]: Reason it works is because `JavadocUtil.isJavadocComment` is smart and do not consider first in line block comment a javadoc.

https://github.com/checkstyle/checkstyle/blob/a1ea36d1db4d78527e7ae2c5d0501c083fa3c586/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/AbstractJavadocCheck.java#L318-L319


https://github.com/checkstyle/checkstyle/blob/a1ea36d1db4d78527e7ae2c5d0501c083fa3c586/src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtil.java#L160-L162


https://github.com/checkstyle/checkstyle/blob/a1ea36d1db4d78527e7ae2c5d0501c083fa3c586/src/main/java/com/puppycrawl/tools/checkstyle/utils/JavadocUtil.java#L340-L356

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/AbstractJavadocCheckTest.java
Comment (romani) [inline src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/AbstractJavadocCheckTest.java:199]: please do not follow our bad pattern of naming with digits.
It was lazy decision that bites after some time.
please make name wordy `testCacheWithTwoBlockCommentAtSameLine`
please do "same" for other test method and Inputs.

**Review by romani — APPROVED** [review]
last before merge.

**Review by romani — CHANGES_REQUESTED** [review]
renaming item above should be fixed.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/abstractjavadoc/InputAbstractJavadocCache4.java
Comment (chris-carneiro) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/javadoc/abstractjavadoc/InputAbstractJavadocCache4.java:14]: @romani okay thanks for the heads up. 
You found the reason for the behavior so It's good news for the current matter. 

I was currently going down the inner workings of the TreeWalker and noticed something similar. 

For future reference I would like point out that the JavaLanguageLexer#reportBlockComment shows the same behavior and caches as Javadoc only the last Javadoc block,
specifically this piece of code:

image
So basically since the map key is the line number, the last comment is overridden by the next comment found on the same line.

Anyways, I thought it was worth pointing out.

File: src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/AbstractJavadocCheckTest.java
Comment (chris-carneiro) [reply src/test/java/com/puppycrawl/tools/checkstyle/checks/javadoc/AbstractJavadocCheckTest.java]: Done

**Comment by chris-carneiro** [comment]
> renaming item above should be fixed.
My bad this PR took so long! Thanks for your help @romani

**Review by romani — APPROVED** [review]
Ok to merge if CI pass

**Comment by romani** [comment]
We have lack of active maintainers now.
So merging on single review.

**Comment by romani** [comment]
I hope @chris-carneiro watched power of pitest to eliminate unnecessary code.

**Comment by chris-carneiro** [comment]
> I hope @chris-carneiro watched power of pitest to eliminate unnecessary code.

I will resolve 1 or 2 more suppressions to feel more confident about it. But It seems quite effective indeed

**Pull request merged**
